cloudconfig/cloudinit/network_ubuntu.go

```diff
@@ -0,0 +1,432 @@
+	// Otherwise we'll use the first device with a gateway address,
+	// it'll be filled in the second loop.
+	for _, info := range interfaces {
+		if info.IsDefaultGateway {
+			switch info.GatewayAddress.Type {
+			case network.IPv4Address:
+				gateway4Address = info.GatewayAddress.Value
+			case network.IPv6Address:
+				gateway6Address = info.GatewayAddress.Value
+			}
+		}
+	}
+
+	for _, info := range interfaces {
+		ifaceName := strings.Replace(info.MACAddress, ":", "_", -1)
+		// prepend eth because .format of python wont like a tag starting with numbers.
+		ifaceName = fmt.Sprintf("{eth%s}", ifaceName)
+
+		if !info.NoAutoStart {
+			autoStarted.Add(ifaceName)
+		}
+
+		if cidr := info.CIDRAddress(); cidr != "" {
+			nameToAddress[ifaceName] = cidr
+		} else if info.ConfigType == network.ConfigDHCP {
+			nameToAddress[ifaceName] = string(network.ConfigDHCP)
+		}
+		nameToRoutes[ifaceName] = info.Routes
+
+		for _, dns := range info.DNSServers {
+			dnsServers.Add(dns.Value)
+		}
+
+		dnsSearchDomains = dnsSearchDomains.Union(set.NewStrings(info.DNSSearchDomains...))
+
+		if info.GatewayAddress.Value != "" {
+			switch {
+			case gateway4Address == "" && info.GatewayAddress.Type == network.IPv4Address:
+				gateway4Address = info.GatewayAddress.Value
+
+			case gateway6Address == "" && info.GatewayAddress.Type == network.IPv6Address:
+				gateway6Address = info.GatewayAddress.Value
+			}
+		}
+
+		if info.MTU != 0 && info.MTU != 1500 {
+			nameToMTU[ifaceName] = info.MTU
+		}
+
+		namesInOrder = append(namesInOrder, ifaceName)
+	}
+
+	prepared := &PreparedConfig{
+		InterfaceNames:   namesInOrder,
+		NameToAddress:    nameToAddress,
+		NameToRoutes:     nameToRoutes,
+		NameToMTU:        nameToMTU,
+		AutoStarted:      autoStarted.SortedValues(),
+		DNSServers:       dnsServers.SortedValues(),
+		DNSSearchDomains: dnsSearchDomains.SortedValues(),
+		Gateway4Address:  gateway4Address,
+		Gateway6Address:  gateway6Address,
+	}
+
+	logger.Debugf("prepared network config for rendering: %+v", prepared)
+	return prepared
+}
+
+// AddNetworkConfig adds configuration scripts for specified interfaces
+// to cloudconfig - using boot textfiles and boot commands. It currently
+// supports e/n/i and netplan.
+func (cfg *ubuntuCloudConfig) AddNetworkConfig(interfaces []network.InterfaceInfo) error {
+	if len(interfaces) != 0 {
+		eni, err := GenerateENITemplate(interfaces)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		netPlan, err := GenerateNetplan(interfaces)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		cfg.AddBootTextFile(jujuNetplanFile, netPlan, 0644)
+		cfg.AddBootTextFile(systemNetworkInterfacesFile+".templ", eni, 0644)
+		cfg.AddBootTextFile(systemNetworkInterfacesFile+".py", NetworkInterfacesScript, 0744)
+		cfg.AddBootCmd(populateNetworkInterfaces(systemNetworkInterfacesFile))
+	}
+	return nil
+}
+
+// Note: we sleep to mitigate against LP #1337873 and LP #1269921.
```

Comment:
LP #1269921 seems still valid (though perhaps not on bionic), but LP #1337873 shouldn't be bothering us anymore.